Support ARM64/AWS Graviton Instances by not performing unaligned accesses

Dillonb opened the pull request:
Resolves https://github.com/Netflix/hollow/issues/517 and allows Hollow to be used on ARM based machines such as AWS Graviton instances.

There are a few things left to finish up, but I wanted to get this in front of people sooner rather than later to get some feedback on the design here.

What I've done is swapped HollowUnsafeHandle from being simply a way to get an instance of sun.misc.Unsafe to a singleton class wrapping sun.misc.Unsafe. It seems the only classes that perform unaligned accesses are the FixedLengthElementArray and SegmentedLongArray families, so it might also make sense to have only those classes call an unaligned helper method on HollowUnsafeHandle, rather than having all unsafe calls go through this class. Let me know if this would be preferred.

I've tested this (but only the consumer side) on an r6g.4xlarge instance on AWS, and performance is very good. Also, the kernel doesn't kill it with a SIGBUS :slightly_smiling_face: 

I'm new to the sun.misc.Unsafe API, so there are a few things I haven't been able to figure out how to do yet:

- How do we detect whether the host platform supports unaligned accesses?

OpenJDK 8 does it like [this internally](https://github.com/openjdk/jdk/blob/9a9add8825a040565051a09010b29b099c2e7d49/jdk/src/share/classes/java/nio/Bits.java#L606-L615), but there seems to not be a public-facing API version of this, at least in JDK8. Maybe a configuration setting would be most appropriate if there's no reliable way to auto detect it.

- Improve efficiency by checking the alignment of the address

I'd like to check if the address is aligned, and use the faster putOrderedLong/getLong calls directly if possible. Since all the accesses use the object + offset form instead of raw addresses, there doesn't seem to be a way to do this unless we can retrieve the address of the base object, which doesn't seem to be possible through sun.misc.Unsafe.

Thanks for your time.

Dillonb committed:
Support unaligned long accesses

Dillonb committed:
remove TODO comments

Sunjeet commented on hollow/src/main/java/com/netflix/hollow/core/memory/HollowUnsafeHandle.java:
```diff
@@ -38,10 +44,102 @@
+    }
+
+    public static HollowUnsafeHandle getUnsafe() {
+        return singleton;
```
We could return an implementation of `HollowUnsafeHandle` that always does unaligned, or an instance that always does aligned. Then we don't incur the cost of branching in the read path.

Dillonb replied:
I'm fairly certain that, since the guard is a final, the JVM will optimize out the branch not-taken when it compiles this code to native, but I can swap to using two implementations to be more certain.

Sunjeet replied:
Ah I didn't consider that.

Sunjeet commented on hollow/src/main/java/com/netflix/hollow/core/memory/HollowUnsafeHandle.java:
```diff
@@ -38,10 +44,102 @@
+    }
+
+    public long getLong(Object o, long l) {
+        if (UNALIGNED_ALLOWED) {
```
I'm curious, if you change this to `if (true)` do you see the SIGBUS? In our testing the only place where we saw SIGBUS was when the delta application code calls `putOrderedLog`. Once we disabled deltas, everything else (initializing from a snapshot, creating indices, un aligned reads, etc.) seemed to work.

Sunjeet replied:
Asking because of performance criticality here

Dillonb replied:
I'll test this and get back to you.

Sunjeet commented on hollow/src/main/java/com/netflix/hollow/core/memory/HollowUnsafeHandle.java:
```diff
@@ -38,10 +44,102 @@
+            return unsafe.getLong(o, l);
+        } else {
+            // Assume little endian
+            return ((long) unsafe.getByte(o, l) & 0xFF)
```
If we end up needing to do this, it will likely be faster to batch these up into- read 2 words when a read spans a word boundary then stitch up the long. 
If I am reading this right the `getByte`s translate into reading word or half word anyway https://developer.arm.com/documentation/ddi0360/f/unaligned-and-mixed-endian-data-access-support/unaligned-access-support/armv6-extensions

Dillonb replied:
Yeah, one of the points I mentioned was detecting when accesses are 1, 2, 4, or 8 bytes aligned and performing the most efficient one (any performance penalty from the branches should be outweighed by the gains from larger reads/writes)

At the time, I wasn't sure how to determine this, since all accesses are done with "object plus offset" methods - but it seems that heap objects are always 8-byte aligned, so we should be able to figure this out based on the offset alone. I'll update the PR with this optimization.

Sunjeet replied:
Heads up the ByteBuffers in `EncodedByteBuffer` and `EncodedLongBuffer` are off heap and won't be word-aligned, I think

Sunjeet commented:
> It seems the only classes that perform unaligned accesses are the FixedLengthElementArray and SegmentedLongArray families

To double check, `EncodedByteBuffer` and `EncodedLongBuffer` also do this

Sunjeet commented:
> It seems the only classes that perform unaligned accesses are the FixedLengthElementArray and SegmentedLongArray families, so it might also make sense to have only those classes call an unaligned helper method on HollowUnsafeHandle, rather than having all unsafe calls go through this class.

That sounds fine. Something else to consider is that we'll at some point need to replace all use of sun.misc.Unsafe with MemorySegment et. al. https://openjdk.java.net/jeps/370

Sunjeet commented:
Hi @Dillonb thanks for looking into this! To me your proposed approach to add ARM support is welcome, it allows us to sidestep the SIGBUS error and incrementally get to a performant and stable Hollow on ARM (note: thats low on the priority list for us at the moment). I think with a little tweak to your pr (make HollowUnsafeHandle polymorphic) we can achieve this without introducing performance regression in the x86/amd64 path.

Also adding @dkoszewnik and @toolbear for visibility, and if they have any concerns or additional perspective on going down this road.

Dillonb commented:
Thanks for the feedback!

Seems the last thing remaining to figure out is how to detect if unaligned accesses are allowed on the host. I'm really hoping to not have to go down the road of string matching on `os.arch`, but it's looking more and more like that's the way forward.

Dillonb committed:
detect alignment and use larger writes if possible

dkoszewnik commented:
This is awesome.  No concerns from me with the approach, and glad to see this being tackled as the need arises.  Thanks a lot for making this happen!

Dillonb committed:
Detect alignment and use larger reads if possible

Sunjeet commented:
@PaulSandoz brought up a couple of interesting points offline, 
* do we really need `putOrderedLong` here, or could we switch to plain store `putLong` that might just work for unaligned access on ARM? 
* this looks a bit cleaner-
```
 if ((offset & 7) == 0) {
        return getLong(o, offset);
    } else if ((offset & 3) == 0) {
        return makeLong(getInt(o, offset),
                        getInt(o, offset + 4));
    } else if ((offset & 1) == 0) {
        return makeLong(getShort(o, offset),
                        getShort(o, offset + 2),
                        getShort(o, offset + 4),
                        getShort(o, offset + 6));
    } else {
        return makeLong(getByte(o, offset),
                        getByte(o, offset + 1),
                        getByte(o, offset + 2),
                        getByte(o, offset + 3),
                        getByte(o, offset + 4),
                        getByte(o, offset + 5),
                        getByte(o, offset + 6),
                        getByte(o, offset + 7));
    }
```

Dillonb commented:
Ah, switching to straight `.putLong` from `.putOrderedLong` actually does resolve the SIGBUS issue. Must be that memory fences must be aligned, but stores can be unaligned. Thanks for the idea!

Dillonb commented:
I've created https://github.com/Netflix/hollow/pull/519 with this new approach.

PaulSandoz commented:
@Sunjeet @Dillonb thanks. I forgot to add that the proposed unaligned ordered put changes the memory order and atomicity guarantees, so it no longer conforms to the semantics of an ordered put (release store). A read can now observe a partial write.

I cannot recall the exact concurrency requirements of these array-like classes, but It's not clear to me why the ordered put (release store) is needed. It's never paired with an acquire load. If not needed i think it greatly simplify things if it can be replaced with a plain atomic store on x64 and ARM, thereby preserving the atomicity guarantees (which i think may be the more important property).

Dillonb commented:
@PaulSandoz just to be absolutely safe, in #519 I've added a storeFence() after every call that I changed from putOrderedLong to putLong, to keep the ordering and atomicity guarantees. Performance seemed to be fine, but I didn't test it on any absurdly large datasets 🙂

Sunjeet commented:
https://github.com/Netflix/hollow/pull/519 was accepted, thanks @Dillonb !
